File: 143-reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseLL(ListNode* head){
        // 2. Reverse Linked List O(n/2) ~ O(n)
        ListNode* second = head;
        ListNode* previous = nullptr;
        while(second != nullptr){
            ListNode* forward = second;
            second = second->next;
            forward->next = previous;
            previous = forward;
        }
        return previous;
    }

    void reorderList(ListNode* head) {
        // Split the linked list in half, reverse the second half, and then merge

        // 1. Get second half O(n)
        ListNode* slow = head;
        ListNode* fast = head->next;
        while((fast != nullptr) && (fast->next != nullptr)){
            slow = slow->next;
            fast = fast->next->next;
        }

        // 3. Merge the two Linked List O(n)
        ListNode* second = reverseLL(slow->next);
        slow->next = nullptr;
        while(second != nullptr){
            ListNode* headNext = head->next;
            ListNode* revNext = second->next;
            head->next = second;
            second->next = headNext;
            head = headNext;
            second = revNext;
        }
    }
};

File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        // find middle
        ListNode slow = head;
        ListNode fast = head.next;
        while((fast != null) && (fast.next != null)){
            slow = slow.next;
            fast = fast.next.next;
        }

        // reverse second half
        // sec, sec.nex, prev
        ListNode second = slow.next;    // slow = 3, sec=4
        ListNode prev = slow.next = null;   
        while(second != null){
            ListNode temp = second.next; //temp=5, temp=n
            second.next = prev; // secnex=null, senx=4
            prev = second;  // prev=4, prev=5
            second = temp;  // sec=5, sec=n
        }   // prev, sec.nex, sec

        // merge two halfs
        ListNode first = head;
        second = prev;
        while(second != null){
            ListNode temp1 = first.next;
            ListNode temp2 = second.next;
            first.next = second;
            second.next = temp1;
            first = temp1;
            second = temp2;
        }
    }
}